idaes/generic_models/unit_models/valve.py

```diff
@@ -0,0 +1,293 @@
+# software owners: The Regents of the University of California, through
+# Lawrence Berkeley National Laboratory,  National Technology & Engineering
+# Solutions of Sandia, LLC, Carnegie Mellon University, West Virginia
+# University Research Corporation, et al. All rights reserved.
+#
+# Please see the files COPYRIGHT.txt and LICENSE.txt for full copyright and
+# license information, respectively. Both files are also available online
+# at the URL "https://github.com/IDAES/idaes-pse".
+##############################################################################
+"""
+This provides standard valve models for adaibatice control valves.  Beyond the
+most common valve models, and adiabatic valve model can be added by supplying
+custom callbacks for the pressure-flow relation or valve function.
+"""
+
+__Author__ = "John Eslick"
+
+from enum import Enum
+
+import pyomo.environ as pyo
+from pyomo.common.config import ConfigValue, In
+
+from idaes.core import declare_process_block_class
+from idaes.generic_models.unit_models.pressure_changer import (
+    PressureChangerData,
+    ThermodynamicAssumption,
+    MaterialBalanceType,
+)
+from idaes.core.util.exceptions import ConfigurationError
+from idaes.core.util import from_json, to_json, StoreSpec
+from idaes.core.util.model_statistics import degrees_of_freedom
+import idaes.logger as idaeslog
+import idaes.core.util.scaling as iscale
+
+_log = idaeslog.getLogger(__name__)
+
+
+class ValveFunctionType(Enum):
+    linear = 1
+    quick_opening = 2
+    equal_percentage = 3
+
+
+def linear_cb(b):
+    """
+    Linear opening valve function callback.
+    """
+    @b.Expression(b.flowsheet().config.time)
+    def valve_function(b2, t):
+        return b2.valve_opening[t]
+
+
+def quick_cb(b, t):
+    """
+    Quick opening valve function callback.
+    """
+    @b.Expression(b.flowsheet().config.time)
+    def valve_function(b2, t):
+        return sqrt(b2.valve_opening[t])
+
+
+def equal_percentage_cb(b):
+    """
+    Equal percentage valve function callback.
+    """
+    b.alpha = pyo.Var(initialize=100, doc="Valve function parameter")
+    b.alpha.fix()
+    @b.Expression(b.flowsheet().config.time)
+    def valve_function(b2, t):
+        return b2.alpha ** (b2.valve_opening[t] - 1)
+
+
+def pressure_flow_default_callback(b):
+    """
+    Add the default pressure flow relation constraint.  This will be used in the
+    valve model, a custom callback is provided.
+    """
+    umeta = b.config.property_package.get_metadata().get_derived_units
+
+    b.Cv = pyo.Var(
+        initialize=0.1,
+        doc="Valve flow coefficent",
+        units=umeta("amount")/umeta("time")/umeta("pressure")**0.5
+    )
+    b.Cv.fix()
+
+    b.flow_var = pyo.Reference(b.control_volume.properties_in[:].flow_mol)
+    b.pressure_flow_equation_scale = lambda x : x**2
+
+    @b.Constraint(b.flowsheet().config.time)
+    def pressure_flow_equation(b2, t):
```

Comment:
Whilst I know why you did it, the use of `b2` looks a little weird.

Reply:
I thought so too, but I couldn't think of anything less weird.  Maybe I'll change the top level b to "valve" or something more descriptive.